source/lac/petsc_precondition.cc

```diff
@@ -1094,43 +1096,106 @@ namespace PETScWrappers
     AssertThrow(ierr == 0, ExcPETScError(ierr));
   }
 
-  int
+#  ifndef PetscCall
+#    define PetscCall(code)             \
+      do                                \
+        {                               \
+          PetscErrorCode ierr = (code); \
+          CHKERRQ(ierr);                \
+        }                               \
+      while (0)
+#  endif
+
+  PetscErrorCode
+  PreconditionShell::pcsetup(PC ppc)
+  {
+    PetscFunctionBeginUser;
+    // Failed reason is not reset uniformly within the
+    // interface code of PCSetUp in PETSc.
+    // We handle it here.
+    PetscCall(PCSetFailedReason(ppc, PC_NOERROR));
+    PetscFunctionReturn(0);
+  }
+
+  PetscErrorCode
   PreconditionShell::pcapply(PC ppc, Vec x, Vec y)
   {
     void *ctx;
 
     PetscFunctionBeginUser;
-    PetscErrorCode ierr = PCShellGetContext(ppc, &ctx);
-    AssertThrow(ierr == 0, ExcPETScError(ierr));
+    PetscCall(PCShellGetContext(ppc, &ctx));
 
     auto user = static_cast<PreconditionShell *>(ctx);
-    AssertThrow(user->vmult,
-                StandardExceptions::ExcFunctionNotProvided(
-                  "std::function vmult"));
+    if (!user->vmult)
+      SETERRQ(
+        PetscObjectComm((PetscObject)ppc),
+        PETSC_ERR_LIB,
+        "Failure in dealii::PETScWrappers::PreconditionShell::pcapply. Missing std::function vmult");
 
     VectorBase src(x);
     VectorBase dst(y);
-    user->vmult(dst, src);
+    const int  lineno = __LINE__;
+    try
+      {
+        user->vmult(dst, src);
+      }
+    catch (const RecoverableUserCallbackError &)
+      {
+        PetscCall(PCSetFailedReason(ppc, PC_SUBPC_ERROR));
+      }
+    catch (...)
+      {
+        return PetscError(
+          PetscObjectComm((PetscObject)ppc),
+          lineno + 3,
+          "vmult",
+          __FILE__,
+          PETSC_ERR_LIB,
+          PETSC_ERROR_INITIAL,
+          "Failure in pcapply from dealii::PETScWrappers::NonlinearSolver");
```

Comment:
Is it correct that by default, `PetscError` simply returns the error code (here `PETSC_ERR_LIB`) and does nothing else? But that one can configure things to print errors on the console with all the information you are giving as arguments?

Reply:
What do you mean? `PetscError` is the entry point for error handling in PETSc, hidden in the macro calls `SETERRQ`, `CHKERRQ` (in the past) and `PetscCall(...)` in the most recent versions. It can start a cascade of errors (when called with `PETSC_ERROR_INITIAL`, used in the `SETERRQ` macro) or continue going up the hierarchy of calls (with `PETSC_ERROR_REPEAT` in the `PetscCall` and `CHKERRQ` ). Inside `PetscError` a few things are done, including the preparation of the line for the traceback and the invocation of the error handler, which by default is `PetscTraceBackErrorHandler`, which prints the stack trace above.  The error handler is customizable by the user.

Reply:
My question was mainly what happens by default when you hit `PetscError`. Does it automatically print a stacktrace, or does it only do that if you configure PETSc to do so (or pass an appropriate command line flag)?

I'm asking because throwing exceptions (recoverable or not) should be a totally legitimate thing to do. The "error" is simply if you don't deal with them somewhere before you are back in `main()`, and at that time the stacktrace should be printed. If you throw and catch an exception, no error should be printed because no error has happened: In the `catch` statement, you are dealing with the error, after all.

Reply:
The default error handling is the one that prints the traceback above. We don't get the `what()` the message when the error is raised by the `solve_with_jacobian` callback because I'm eating the excpetion in `PetscPreconditioner`. Before over-engineering the code, we should probably decide what to do in terms of unifying the usage of `AssertPETSc` (i.e. throw exception, i.e. when calling PETSc in deal.II world), `PetscCall` (i.e. calling PETSc in PETSc callbacks like `snes_function` in `NonlinearSolver`) and, possibly, have a single and unified `call_and_possibly_recover` that will be shared by all the PETSc code in deal.II

Reply:
@bangerth On second thought (and by looking at https://github.com/dealii/dealii/pull/15349 and https://github.com/dealii/dealii/pull/15360) I think I should handle the exception here to have a more uniform error report from the nonlinear solvers. I will push one last change before merging this

Reply:
Too late -- the PR is already merged :-)